google-cloud-spanner/lib/google/cloud/spanner/interval.rb

```diff
@@ -0,0 +1,261 @@
+          end
+
+          # Returns an Interval instance with the given microseconds.
+          #
+          # @param [Integer]
+          # @return [Interval]
+          def from_microseconds microseconds
+            nanoseconds = microseconds * NANOSECONDS_IN_A_MICROSECOND
+            Interval.new 0, 0, nanoseconds
+          end
+
+          # Returns an Interval instance with the given nanoseconds.
+          #
+          # @param [Integer]
+          # @return [Interval]
+          def from_nanoseconds nanoseconds
+            Interval.new 0, 0, nanoseconds
+          end
+
+          private
+
+          def years_to_months years
+            Integer(years) * 12
+          end
+
+          def hours_to_nanoseconds hours
+            Integer(hours) * NANOSECONDS_IN_AN_HOUR
+          end
+
+          def minutes_to_nanoseconds minutes
+            Integer(minutes) * NANOSECONDS_IN_A_MINUTE
+          end
+
+          def seconds_to_nanoseconds seconds
+            # We only support up to nanoseconds of precision
+            split_seconds = seconds.to_s.split '.'
+            if split_seconds.length > 2 || split_seconds.length > 1 && split_seconds[1].length > 9
+              raise ArgumentError, "The seconds portion of the interval only supports up to nanoseconds."
+            end
+
+            Float(seconds) * NANOSECONDS_IN_A_SECOND
+          end
+        end
+
+        def to_s
+          # Memoizing it as the logic can be a bit heavy
+          @string_representation ||= self.to_string
+        end
+
+        private
+
+        def initialize months, days, nanoseconds
+          if (months > MAX_MONTHS || months < MIN_MONTHS)
+            raise ArgumentError, "The Interval class supports months from #{MIN_MONTHS} to #{MAX_MONTHS}."
+          end
+          @months = months
+
+          if (days > MAX_DAYS || days < MIN_DAYS)
+            raise ArgumentError, "The Interval class supports days from #{MIN_DAYS} to #{MAX_DAYS}."
+          end
+          @days = days
+
+          if (nanoseconds > MAX_NANOSECONDS || nanoseconds < MIN_NANOSECONDS)
+            raise ArgumentError, "The Interval class supports nanoseconds from #{MIN_NANOSECONDS} to #{MAX_NANOSECONDS}"
+          end
+          @nanoseconds = nanoseconds
+        end
+
+        def to_string
+          years = 0
+          months = 0
+          days = @days
+          hours = 0
+          minutes = 0
+          seconds = 0
+          remaining_nanoseconds = @nanoseconds
+
+          years = @months / 12
+          months = @months % 12
+          hours = Integer(remaining_nanoseconds / 3_600_000_000_000)
+          remaining_nanoseconds %= 3_600_000_000_000
+          minutes = Integer(remaining_nanoseconds / 60_000_000_000)
+          remaining_nanoseconds %= 60_000_000_000
+          seconds = remaining_nanoseconds / 1_000_000_000
+
+          interval_string = "P"
+
+          if years != 0
+            interval_string += "#{years}Y"
```

Comment:
Recommend against appending to strings. Instead, append to an array of strings, and then use Array#join at the end.